### ISSUE TITLE
Do image captions have any effect during training?

### ISSUE THREAD

**Issue by FurkanGozukara** [rinongal/textual_inversion#131]
During training do image captions have any importance?

Or the training is done solely based on training dataset and no captions prompts other text are used at all?

**Comment by FurkanGozukara** [comment]
@rinongal @tmuskal @ryanrussell @0x2b3bfa0 

Why I am asking? Because I am making tutorial guides videos for Stable Diffusion

My latest video is about Textual Inversion
I am also trying to give as much as possibly technical correct information

All of the current tutorials are incorrect if no image captions are utilized at all during the training

I also read the paper used a lot info from there but couldnt figure out this part

my latest video : [How To Do Stable Diffusion Textual Inversion (TI) / Text Embeddings By Automatic1111 Web UI Tutorial](https://youtu.be/dNOpWt-epdQ)

**Comment by Shondoit** [comment]
Captions are definitely used during training (as long as your templates contain [filewords])

They allow you to describe what you _don't_ want the embedding to learn from the training image. And they create relationships between the embedding and prompts.

1.You can consider the caption (a.k.a. training prompt) to be the formula of the training image. I.e. what do I need to write in the prompt to get this image back.

Now, the prompt will contain your TI embedding, that is flexible. The other words are fixed.  When you do training, you change the embedding to better match the outcome.

When you have a picture of a person wearing a sombrero in front of a palm tree, you could think of the prompt as "person + sombrero + palm tree = training image"
If you're trying to train the person, then your prompt would be "[name] + sombrero + palm tree = training image"
If you solve for name: [name] = training image - sombrero - palm tree (i.e. your embedding would be of the image, but ignore the things you described.

Another way to look at it: since you've already described the other objects in the prompt, leaving out the embedding would already draw the tree and sombrero. So adding it, it would have no reason to learn it too. The drawing of those objects has been delegated to the prompt.

2.It creates relationships.
If you train a person with sometimes glasses, sometimes not, one picture with blue hair, 3 pictures with their eyes closed,  it's may learn a blend of everything, leading to weird half closed eyes with weird outlines (from the glasses)
Describing the things that are unusual for this person can separate the concepts and serve as relationships for later.
If you've trained with captions and later you generate a drawing "person with blue hair wearing glasses" it can reference what it learned during training.

**Comment by Shondoit** [comment]
You can try some experiments if you want.

Caption all pictures from the left as "[name] wearing a hat" or some other silly nonsense.
All pictures from the right leave uncaptioned.
I wonder if you can consistently generate the person shown from the left by reusing that caption as prompt.

I honestly don't know, I'm curious about it

**Comment by FurkanGozukara** [comment]
> Captions are definitely used during training (as long as your templates contain [filewords])
> 
> They allow you to describe what you _don't_ want the embedding to learn from the training image. And they create relationships between the embedding and prompts.
> 
> 1.You can consider the caption (a.k.a. training prompt) to be the formula of the training image. I.e. what do I need to write in the prompt to get this image back.
> 
> Now, the prompt will contain your TI embedding, that is flexible. The other words are fixed. When you do training, you change the embedding to better match the outcome.
> 
> When you have a picture of a person wearing a sombrero in front of a palm tree, you could think of the prompt as "person + sombrero + palm tree = training image" If you're trying to train the person, then your prompt would be "[name] + sombrero + palm tree = training image" If you solve for name: [name] = training image - sombrero - palm tree (i.e. your embedding would be of the image, but ignore the things you described.
> 
> Another way to look at it: since you've already described the other objects in the prompt, leaving out the embedding would already draw the tree and sombrero. So adding it, it would have no reason to learn it too. The drawing of those objects has been delegated to the prompt.
> 
> 2.It creates relationships. If you train a person with sometimes glasses, sometimes not, one picture with blue hair, 3 pictures with their eyes closed, it's may learn a blend of everything, leading to weird half closed eyes with weird outlines (from the glasses) Describing the things that are unusual for this person can separate the concepts and serve as relationships for later. If you've trained with captions and later you generate a drawing "person with blue hair wearing glasses" it can reference what it learned during training.

actually this is what i also assumed and explained in the video

however from discord Hackmans#6743 said the otherwise

i explained in the video as, vectors of those prompt tokens are used to reach your target subject image. therefore, lets say you used eyeglasses as token, the model will not be have to learn your eyeglasses since already using a helper vector for that

this is a good explanation right?

**Comment by Shondoit** [comment]
Yes that a good explanation.  (Although I would say that the embedding is learning, not the model)

**Comment by FurkanGozukara** [comment]
> Yes that a good explanation. (Although I would say that the embedding is learning, not the model)

ye I think i said that way not the model :D thank you for clarification

**Comment by rinongal** [comment]
Sorry for the late response. @Shondoit 's answer is more or less spot on.

The templates are indeed used in training, and you can use them to help direct the network towards what you want to learn by describing irrelevant parts of the image.

Maybe one thing I'll add here is that they also help ensure that the model knows the new embeddings should 'play nicely' with many different prompts and when the word is inserted in different indices in the text.

**Comment by FurkanGozukara** [comment]
> Sorry for the late response. @Shondoit 's answer is more or less spot on.
> 
> The templates are indeed used in training, and you can use them to help direct the network towards what you want to learn by describing irrelevant parts of the image.
> 
> Maybe one thing I'll add here is that they also help ensure that the model knows the new embeddings should 'play nicely' with many different prompts and when the word is inserted in different indices in the text.

great info. i made a tutorial on this and explained this way
i will make a pull request to add the readme. i hope you accept it
tutorial link : https://youtu.be/dNOpWt-epdQ 

we even provided example training dataset

**Comment by GeneralAwareness** [comment]
If I have a picture of a robot and I only want it to learn the robot then in all the caption post BliP I go in and remove the reference to the robot?  I tried this for a dismal failure.  Matter of a fact I despise captions so I try NOT to use them.

A simple example was I went in and cut out the only person in each of the photos.  What was left was a black silhouette of a man (to make sure I used CLIP V2).  GREAT, makes it easy as I need only caption it as that YET it learned that and screwed everything up to the point I have to question if it even works.

**Comment by jndietz** [comment]
Can we caption just by running `main.py`?
